pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java

```diff
@@ -1082,7 +1087,16 @@ public void start() throws PulsarServerException {
     }
 
     public void runWhenReadyForIncomingRequests(Runnable runnable) {
-        readyForIncomingRequestsFuture.thenRun(runnable);
+        // Here we don't call the thenRun() methods because CompletableFuture maintains a stack for pending callbacks,
+        // not a queue. Once the future is complete, the pending callbacks will be executed in reverse order of
+        // when they were added.
+        synchronized (pendingTasksBeforeReadyForIncomingRequests) {
+            if (readyForIncomingRequestsFuture.isDone()) {
+                runnable.run();
+            } else {
```

Comment:
This requires a slight change so that it doesn't synchronize the running of the runnable needlessly after the broker is initialized. One possible solution is to set a flag and if the flag is set, to run the runnable outside of the synchronized block.

Reply:
Good suggestion. I will add a flag

Reply:
Addressed, PTAL again.

Reply:
Looks good